File: partitionLabels_763.py
# // Time Complexity :  O(n)
# // Space Complexity : O(n), hashmap unless we specify space of hashmap to be of 26 length as only lowercase letters are required.
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

# // Your code here along with comments explaining your approach

#Approach and intuition:

#Brute Force: Form all the partitions you can and check if the partitions are valid or not. Exponential runtime.

# Optimal: What if we keep track of last occurence of the character? Problems: What if the first char is not the last char of the partition?
# For every element, we'll keep the furthest away point(since we don't know the furthest away element).
#Eg: "bababcbacadefegdehijhklij"
#Create hashmap: {a:9, b:6, c:8, d:15, e:16, f:12, g:14, h:17, i:23,j:24,k:21}
# Two pointers: start, end-> defines partition and i iterates through the loop.
# @i: check if S[i] is the furthest away point for that letter? No? Move end till furthest away index(from hmap). move i ahead, is S[i]'s furthest away index still greater than the end index? No? Move end to new furthest away index, else do nothing.
# If i==end: add to o/p= end-start+1, start= i+1

class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        if S == None or len(S)==0:
            return []
        
        hmap = {}
        
        for i in range(len(S)):
            ch = S[i]   
            hmap[ch] = i
        
        start, end = 0,0
        output = []
        #iterate through string and calc o/p
        for i in range(len(S)):
            if end<hmap[S[i]]:
                end = hmap[S[i]]
                
            if i == end:
                output.append(end-start+1)
                start = i+1
                
        return output

File: Partition_Labels_763.java
//Time Complexity : O(n)
//Space Complexity : O(1)
//Did this code successfully run on Leetcode : Yes
//Any problem you faced while coding this : None

package com.s30.satish;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

class Partition_Labels_763 {
    public List<Integer> partitionLabels(String S) {
        List<Integer> list = new ArrayList<>();
        if(S == null || S.length() == 0)
            return list;
        int start = 0;
        int end = 0;
        HashMap<Character, Integer> map = new HashMap<>();
        for(int i = 0; i < S.length(); i++)
        {
            map.put(S.charAt(i), i);
        }
        for(int i = 0; i < S.length(); i++)
        {
            char c = S.charAt(i);
            end = Math.max(map.get(c), end);
            if(i == end)
            {
                list.add(end - start + 1);
                start = i+1;
            }
        }
        return list;
        
//        // nArray approach uses less execution time
//        public List<Integer> partitionLabels(String S) {
//            List<Integer> list = new ArrayList<>();
//            if(S == null || S.length() == 0)
//                return list;
//            int start = 0;
//            int end = 0;
//            int[] nArray = new int[26];
//            for(int i = 0; i < S.length(); i++)
//            {
//                nArray[S.charAt(i) - 'a'] = i;
//            }
//            for(int i = 0; i < S.length(); i++)
//            {
//                char c = S.charAt(i);
//                end = Math.max(nArray[c - 'a'], end);
//                if(i == end)
//                {
//                    list.add(end - start + 1);
//                    start = i+1;
//                }
//            }
//            return list;
//        }
        
        
    }
    public static void main(String[] args)
    {
    	Partition_Labels_763 obj = new Partition_Labels_763();
    	String str = "ababcbacadefegdehijhklij";
    	System.out.print(obj.partitionLabels(str));
    	
    }
}
